manual/modshell.1
.TH ModShell 1 "2024 Szeptember 30" ""
.SH NAME
modshell \- Command-driven scriptable Modbus utility
.SH SYNOPSIS
.B modshell
[options]
.SH COPYRIGHT
ModShell is Copyright (C) 2023-2025 by Pozsar Zsolt.
.SH DESCRIPTION
This is a utility that can be used on several operating systems, which
can communicate with connected equipment using Modbus/ASCII, Modbus/RTU
and Modbus/TCP protocols. The basic communication protocol of the program
is Modbus, but DCON and HART was also implemented due to communication with
other devices. The range of foreign protocols may be expanded later.
The program is designed for DOS, Windows and Unix-like operating systems.
.SH OPTIONS
.TP
.B \-f \-\-fullscreen
Start in full screen mode.
.TP
.B \-h \-\-help
Write useable parameters.
.TP
.B \-r \-\-run script
Run a script.
.TP
.B \-v \-\-version
Write version and build information.
.SH OPERATION
The program can - even automatically - read, write or copy data from one device
to another (e.g. transferring settings). When copying, the source and
destination register areas can be different.
.PP
The program has a traditional (CLI) or full-screen (TUI) command-line interface
and is also suitable for running pre-created scripts independently (as a command
interpreter). The program provides help on the commands that can be used, and
offers possible values when the parameters are entered incorrectly. The issued
commands are placed in history, which can be browsed with the up/down arrow keys.
.SS Operating principle
It must be defined the I/O devices, then the protocols and the connections.
There can be eight of each. The data traffic takes place between the preset
connections. In all cases, the data is sent to or read from the internal buffer.
The size of the buffer is suitable for storing 2*9999 logical and word values of
the same size.
.PP
One hundred and twenty-eight variables or constants can be created
in the program, to which we can assign a value of any type (eg.: message, boolean
or integer register value, etc.) Variables and constants can be used to perform
logical and basic mathematical operations, and can be used to pass values to commands.
.PP
It is possible to create 32 single-line macros to replace frequently used longer
commands. The program includes 3 pre-created macros, the content of which can
also be changed.
.SS Projects
In the program, you can create projects for easier management of settings and
data. The name of the current project is shown in the prompt. The project
directory will be created in the program directory on DOS, and in the ModShell
directory in the user's home directory on all other systems. If only filename
is specified during file operations (without path), this directory will be the
source/destination directory.
.SS File operations
The command line history can be exported to a text file and provides it with the
appropriate 'shebang' for all operating system. You can easily create a script
from this raw file.
.PP
Device, protocol and connection settings can be saved and loaded in their own
format. During saving, three typed files are created, with the following
extensions: DDT, PDT, CDT.
.PP
All register values can be saved and loaded in their own format. During saving,
two typed files are created, with the following extensions: BDT, WDT.
.PP
One or more same type registers can be exported to file. During saving, one text
file is created, with CSV, INI or XML extension. The program can only import from
INI and XML format files.
.PP
The program has simple serial, TCP and UDP consoles whose traffic is logged in
the project directory as console.log.
.PP
On exit, the command line history, echo, input and send method and colors are
preserved.
.PP
The program also has basic file and directory management commands.
.SS Script operations
The script on the disc can be loaded into the already running program and started
manually.The loaded script can be edited with a line editor, saved to disk or
deleted from the buffer. The graphical version has a simple editor window
with syntax highlighting instead of the line editor.
Variables and constants defined before running the script will be deleted.
It is also possible to observe the values of up to four variables during
runtime and to keep the final values​of constants and variables created
during runtime.
In interpreter mode, parameters are passed from the OS command line to the script
and the script's exit value is returned to the OS. This parameters are available
as predefined constants.
.SS Serial console
The program also provides the possibility to send and receive raw data
via a serial port, and also includes a very simple serial console. The
display of sent and received data can be turned off or raw text and
hexadecimal viewing can be selected.
.SS TCP console
The program also provides the possibility to send and receive raw data
via network device with TCP, and also includes a very simple TCP console. The
display of sent and received data can be turned off or raw text and
hexadecimal viewing can be selected.
.SS UDP console
The program also provides the possibility to send and receive raw data
via network device with UDP, and also includes a very simple UDP console. The
display of sent and received data can be turned off or raw text and
hexadecimal viewing can be selected.
.SS Direct I/O access
The program provides direct, byte-sized reading and writing of hardware I/O
ports on PCs and bit-sized reading and writing GPIO ports on Raspberry Pi and
Industrial PCs. On Windows operating systems, it uses the included external
dynamically linked library.   

.SH PREDEFINED VARIABLES
.TP
\fB$verbose\fP
Verbosity level (values: NOTHING|ERROR|ALL)
.TP
Empty and any other variable value are equivalent to ALL.

.SH PREDEFINED CONSTANTS
.TP
\fB$?\fP
Exit value of the commands.
.TP
\fB$ARGx\fP
OS command line arguments (interpreter mode).
.TP
\fB$ARGCNT\fP
Number of the OS command line arguments (interpreter mode).
.TP
\fB$B01\fP
150 (baud).
.TP
\fB$B03\fP
300 (baud).
.TP
\fB$B06\fP
600 (baud).
.TP
\fB$B1\fP
1200 (baud).
.TP
\fB$B2\fP
2400 (baud).
.TP
\fB$B4\fP
4800 (baud).
.TP
\fB$B8\fP
9600 (baud).
.TP
\fB$B19\fP
19200 (baud).
.TP
\fB$B38\fP
38400 (baud).
.TP
\fB$B57\fP
57600 (baud).
.TP
\fB$B115\fP
115200 (baud).
.TP
\fB$EULER\fP
Value of e  (2.7182818284590452354).
.TP
\fB$HOME\fP
User's home directory.
.TP
\fB$PI\fP
Value of Pi (3.1415926535897932385).
.TP
\fB$PRJDIR\fP
Directory of the actual project.
.TP
\fB$PRJNAME\fP
Name of the actual project.
.TP
\fB$SQRT2\fP
Value of square root of 2.
.TP
\fB$SQRT3\fP
Value of square root of 3.

.SH PREDEFINED MACROS
.TP
\fBmydev0\fP
Set dev0 set /dev/ttyUSB0 9600 8 n 1.
\fB|mypro0\fP
Set pro0 rtu 1.
\fB|mycon0\fP
Set con0 dev0 pro0.
.TP
These macros can also be changed.

.SH BUILTIN COMMANDS
Notes:
  - Register is the local buffer register, and remote register is the register of the connected device.
  - The '$' sign indicates value of a variable or constant not a direct value.
  - The symbol '[x]' indicates the element number of a variable or constant array.
  - The exit value of the commands is in the '$?' constant.
  - Enclose the string in '"' characters, use '\' before a space, and use '^[' in ANSI sequences.
  

.SS Arithmetical commands
.TP
\fBabs\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates absolute value of the \fI[$]VALUE\fP and the result is placed in
\fI$TARGET\fP.
.TP
\fBadd\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Adds the two values or variables (\fI[$]VALUE1\fP \fI[$]VALUE2\fP)
and the result is placed in \fI$TARGET\fP.
.TP
\fBavg\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP [\fI[$]VALUE3...6\fP]
Calculates average of 2-5 numbers and the result is placed in \fI$TARGET\fP.
.TP
\fBconv\fP \fI$TARGET\fP \fI$FROM\fP|\fIbin\fP|\fIdec\fP|\fIhex\fP|\fIoct\fP \fI$TO\fP|\fIbin\fP|\fIdec\fP|\fIhex\fP|\fIoct\fP \fI[$]VALUE\fP
Converts numbers between binary, decimal, hexadecimal and octal numeral system. Result is placed in \fI$TARGET\fP variable.
\fIVALUE\fP must be between 0 and 65535
.TP
\fBcos\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates cosine of \fI[$]VALUE\fP and
the result is placed in \fI$TARGET\fP.
.TP
\fBcotan\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates cotangent of \fI[$]VALUE\fP
and the result is placed in \fI$TARGET\fP.
.TP
\fBdec\fP \fI$VARIABLE\fP
Decrements \fI$VARIABLE\fP integer.
.TP
\fBdiv\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Divides \fI[$]VALUE1\fP by \fI[$]VALUE2\fP and the result
is placed in \fI$TARGET\fP.
.TP
\fBexp\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates natural exponential of \fI[$]VALUE\fP
and the result is placed in \fI$TARGET\fP.
.TP
\fBidiv\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Divides \fI[$]VALUE1\fP by \fI[$]VALUE2\fP and the result
is placed in \fI$TARGET\fP. (integer division)
.TP
\fBimod\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Divides \fI[$]VALUE1\fP by \fI[$]VALUE2\fP and the result
is placed in \fI$TARGET\fP. (modulus division)
.TP
\fBinc\fP \fI$VARIABLE\fP
Increments \fI$VARIABLE\fP integer.
.TP
\fBln\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates natural logarithm of \fI[$]VALUE\fP
and the result is placed in \fI$TARGET\fP.
.TP
\fBmul\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Multiplies the two two values or variables (\fI[$]VALUE1\fP \fI[$]VALUE2\fP)
and the result is placed in \fI$TARGET\fP.
.TP
\fBmulinv\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates multiplicative inverse of \fI[$]VALUE\fP
and the result is placed in \fI$TARGET\fP.
.TP
\fBodd\fP \fI$TARGET\fP \fI[$]VALUE\fP
Decides whether the integer \fI[$]VALUE\fP is even or odd.
.TP
\fBpow\fP \fI$TARGET\fP \fI[$]BASE\fP \fI[$]EXPONENT\fP
Calculates \fI[$]EXPONENT\fP exponentiation of the \fI[$]BASE\fP
and the result is placed in \fI$TARGET\fP.
.TP
\fBpow2\fP \fI$TARGET\fP \fI[$]EXPONENT\fP
Calculates \fI[$]EXPONENT\fP exponentiation of two
and the result is placed in \fI$TARGET\fP.
.TP
\fBprop\fP \fI$TARGET\fP \fI[$]MIN\fP \fI[$]MAX\fP \fI[$]ZERO\fP \fI[$]SPAN\fP \fI[$]VALUE\fP
Propotional value calculation (for example: 4-20 mA current loop and measured value).
.TP
\fBrnd\fP \fI$TARGET\fP \fI[$]VALUE\fP
Makes random integer number between 0 and
\fI[$]VALUE\fP and the result
is placed in \fI$TARGET\fP.
.TP
\fBround\fP \fI$TARGET\fP \fI[$]VALUE\fP \fI[$]DEC_PLACES\fP
Rounds value or variable (\fI[$]VALUE\fP) to \fI[$]DEC_PLACES\fP decimal places
and the result is placed in \fI$TARGET\fP.
.TP
\fBsin\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates sine of \fI[$]VALUE\fP and the
result is placed in \fI$TARGET\fP.
.TP
\fBsqr\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates square of \fI[$]VALUE\fP and
the result is placed in \fI$TARGET\fP.
.TP
\fBsqrt\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates square root of \fI[$]VALUE\fP and
the result is placed in \fI$TARGET\fP.
.TP
\fBsub\fP \fI$TARGET\fP \fI[$]VALUE1\fP \fI[$]VALUE2\fP
Substracts \fI[$]VALUE2\fP from \fI[$]VALUE1\fP and the result
is placed in \fI$TARGET\fP.
.TP
\fBswp\fP \fI$VARIABLE1\fP \fI$VARIABLE2\fP
Swaps value of the \fI$VARIABLE1\fP and \fI$VARIABLE2\fP.
.TP
\fBtan\fP \fI$TARGET\fP \fI[$]VALUE\fP
Calculates tangent of \fI[$]VALUE\fP and
the result is placed in \fI$TARGET\fP.

.SS Communication commands
.TP
\fBcopyreg\fP \fIcon?\fP \fI$REGTYPE\fP|\fIdinp\fP|\fIcoil\fP \fIcon?\fP \fI$REGTYPE\fP|\fIcoil?\fP \fI[$]ADDRESS\fP [\fI[$]COUNT\fP]
Copies logical data between connections (\fIcon?\fP). Number of connection
(\fIcon?\fP) must be between 0-7, register start \fIADDRESS\fP and the
\fICOUNT\fP must be between 0-9998.
.TP
\fBcopyreg\fP \fIcon?\fP \fI$REGTYPE\fP|\fIireg\fP|\fIhreg\fP \fIcon?\fP \fI$REGTYPE\fP|\fIhreg?\fP \fI[$]ADDRESS\fP [\fI[$]COUNT\fP]
Copies numeral data between connections (\fIcon?\fP). Number of connection
(\fIcon?\fP) must be between 0-7, register start \fIADDRESS\fP and the
\fICOUNT\fP must be between 0-9998.
.TP
\fBdcon\fP \fIcon?\fP \fI$TXARRAY\fP \fI$RXARRAY\fP
Reads or writes data from/to remote device with DCON protocol. The command
takes the input data from the \fI$TXARRAY\fP array, the output data is placed in
the \fI$RXARRAY\fP. Both arrays can only be variable arrays. If necessary, arrays
are resized by the command.
.EX

\fI$$TXARRAY\fP:
  [0]  delimiter [$, #, %, @]
  [1]  command
  [2]  input data
  [3]  checksum enable/disable [1/0]

\fI$$RXARRAY\fP:
  [0]  delimiter [!, ?, >]
  [1]  command (non used)
  [2]  output data

document/changelog
modshell (0.1-beta2) unstable; urgency=low

  * Second user test release will be with the following changes:
  * New command `abs`;  
  * new command `input`;  
  * new commands `pipe`, `stack`;  
  * new command `swp`;  
  * testscript directory for testing commands;  
  * compressed HTML (CHM) help (only XModShell);  
  * (...)
  * syntax highlighter and suggestion extension for Microsoft Visual Studio Code.

 -- Pozsár Zsolt <pozsarzs@gmail.com> dow, dd Mmm 2025 00:00:00

modshell (0.1-beta1) unstable; urgency=low

  * First user test release will be with the following changes:
  * Elimination of confusion between the terms 'register number' and 'data
    address';
  * in interpreter mode, passing parameters from the OS command line to the
    script and returning the script's exit value to the OS;
  * in interpreter mode, command verbosity level setting with a predefined
    variable;
  * ANSI escape sequences support (only DOS and *nix like systems);
  * command `color` -> `set color`, sets all default colors (CLI and TUI);  
  * command `echo` -> `echometh`, parameters: off|on|hex|swap -> off|an|hex|swap;
  * commands `sercons`, `tcpcons`, `udpcons`: character-to-character or string
    sending, with alphanumerical or hexadecimal input, with or without
    alphanumerical or hexhexadecimal echo;  
  * commands `serwrite`, `tcpwrite`, `udpwrite`: alphanumerical or hexadecimal
    input, with or without alphanumerical or hexadecimal echo;
  * DCON protocol support;  
  * keywords (coil, hreg, asc, hex, etc.) should also be specified from
    variables;
  * main menu for all consoles in GUI version;  
  * Modbus register number/address converter utility (in your own scripting
    language);
  * Number converter utility (in your own scripting language);
  * modified source code structure of XModshell;  
  * new command `chkdevlock`/`rmdevlock` (only *nix systems);  
  * new command `datatype`;  
  * new command `exist`;  
  * new command `inputmeth`;  
  * new command `macro`;
  * new command `mbconv`;  
  * new command `mbmon`;  
  * new command `printcolor`(only CLI and TUI);  
  * new command `runmeth`;  
  * new command `sendmeth`;  
  * new command `tcpcons`, `tcpread`, `tcpwrite`;  
  * new command `udpcons`, `udpread`, `udpwrite`;  
  * new commands `ioread` and `iowrite`;
  * new menu items in the main menu for quick execution of Modbus R/W commands;  
  * new menu items for show and edit all register's value in a big table;  
  * new serial baudrates: 150, 300, 600 baud;  
  * serial echo server utility for testing connectivity;  
  * serial ModBus traffic monitor utility;
  * support for variable and constant arrays;  
  * new predefined constants;  
  * syntax highlighter file for applications using GTKSourceView (for example:
    Builder, Geany, Gedit, Mousepad, Pluma, Scribes), for Notepad++ (Windows
    only) and Vim/Neovim;
  * TCP and UPD echo server utilities for testing connectivity.
 
 -- Pozsár Zsolt <pozsarzs@gmail.com> Wed, 01 Jan 2025 00:00:00

modshell (0.1-alpha3) unstable; urgency=low

  * Next developer test release.
  * New source code directory structure;
  * new commands;
  * new predefined constants;
  * bug fixes;
  * GUI (except DOS);
  * script syntax highlighting file for MCEdit and Nano.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Sun, 20 Oct 2024 09:37:44
 
modshell (0.1-alpha2) unstable; urgency=low

  * Next developer test release.
  * Bug fixes;
  * Modbus/ASCII and Modbus/RTU communication;
  * handling of constants;
  * three predefined constants;
  * 28 new command (total: 94);
  * script syntax highlighting file for Micro.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Mon, 12 Feb 2024 00:00:00

modshell (0.1-alpha1) unstable; urgency=low

  * Initial release.

 -- Pozsár Zsolt <pozsarzs@gmail.com> Fri, 16 Jan 2024 21:12:00


source/modshell/modshell.pas
{ +--------------------------------------------------------------------------+ }
{ | ModShell v0.1 * Command-driven scriptable Modbus utility                 | }
{ | Copyright (C) 2023-2025 Pozsar Zsolt <pozsarzs@gmail.com>                | }
{ | modshell.pas                                                             | }
{ | main program                                                             | }
{ +--------------------------------------------------------------------------+ }
{ 
  This program is free software: you can redistribute it and/or modify it
  under the terms of the European Union Public License 1.2 version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.
}
{ 
  run methode #0: simple command line
              #1: show useable arguments
              #2: show version and build information
              #3: full sceen command line
              #4: interpreter mode
              #5: GUI mode
}

{$IFDEF WINDOWS} {$APPTYPE CONSOLE} {$ENDIF}
{$MODE OBJFPC} {$H+} {$MACRO ON}

program modshell;
uses
  {$IFDEF GO32V2} protcom, {$ELSE} Synaser, {$ENDIF}
  {$IFDEF GO32V2} {$ELSE} BlckSock, {$ENDIF}
  {$IFDEF UNIX} Ports, {$ENDIF}
  {$IFDEF WINDOWS} Windows, {$ENDIF}
  Classes,
  DOM,
  GetText,
  INIFiles,
  Math,
  Strings,
  SysUtils,
  XMLRead,
  XMLWrite,
  convert,
  crt,
  dos,
  ucommon,
  uconfig,
  utranslt;

{$I type.pas}
{$I const.pas}
{$I var.pas}
{$I define.pas}

{$IFNDEF GO32V2} {$R *.res} {$ENDIF}

{$I resstrng.pas}

function boolisitconstantarray(s: string): boolean; forward;
function boolisitconstant(s: string): boolean; forward;
function boolisitvariablearray(s: string): boolean; forward;
function boolisitvariable(s: string): boolean; forward;
function boolvalidconstantarraycell(s: string): boolean; forward;
function boolvalidvariablearraycell(s: string): boolean; forward;
function cmd_set(p1, p2, p3, p4, p5, p6, p7: string): byte; forward;
function cmd_run(p1, p2: string): byte; forward;
function intisitconstantarrayelement(s: string): integer; forward;
function intisitconstantarray(s: string): integer; forward;
function intisitconstant(s: string): integer; forward;
function intisitvariablearrayelement(s: string): integer; forward;
function intisitvariablearray(s: string): integer; forward;
function intisitvariable(s: string): integer; forward;
function isitconstantarray(s: string): string; forward;
function isitconstant(s: string): string; forward;
function isitmessage(s: string): string; forward;
function isitvariablearray(s: string): string; forward;
function isitvariable(s: string): string; forward;
function parsingcommands(command: string): byte; forward;
procedure clearallconstants; forward;
procedure clearallmacros; forward;
procedure clearallvariables; forward;
procedure interpreter(f: string); forward;
procedure version(h: boolean); forward;

{$IFDEF UNIX}
  function ioperm(from: Cardinal; num: Cardinal; turn_on: Integer): Integer;
           cdecl; external 'libc';
{$ENDIF}

{$I dll.pas}
{$I lockfile.pas}
{$I validity.pas}
{$I vrbosity.pas}

{$I io.pas}
{$I gpio.pas}
{$I network.pas}
{$I serport.pas}

{$I dcon.pas}
{$I hart.pas}
{$I mbascii.pas}
{$I mbrtu.pas}
{$I mbtcp.pas}
{$I mbutil.pas}
{$I modbus.pas}

{$I cmd_aplg.pas}
{$I cmd_arr.pas}
{$I cmd_asci.pas}
{$I cmd_avg.pas}
{$I cmd_carr.pas}
{$I cmd_cons.pas}
{$I cmd_conv.pas}
{$I cmd_copy.pas}
{$I cmd_cron.pas}
{$I cmd_date.pas}
{$I cmd_dcon.pas}
{$I cmd_dttp.pas}
{$I cmd_dump.pas}
{$I cmd_edit.pas}
{$I cmd_ehmt.pas}
{$I cmd_eras.pas}
{$I cmd_exph.pas}
{$I cmd_expr.pas}
{$I cmd_for.pas}
{$I cmd_get.pas}
{$I cmd_goto.pas}
{$I cmd_gpio.pas}
{$I cmd_gw.pas}
{$I cmd_hart.pas}
{$I cmd_help.pas}
{$I cmd_if.pas}
{$I cmd_impr.pas}
{$I cmd_inmt.pas}
{$I cmd_inpt.pas}
{$I cmd_io.pas}
{$I cmd_labl.pas}
{$I cmd_lcfg.pas}
{$I cmd_let.pas}
{$I cmd_list.pas}
{$I cmd_lock.pas}
{$I cmd_logc.pas}
{$I cmd_lreg.pas}
{$I cmd_lscr.pas}
{$I cmd_math.pas}
{$I cmd_mbcv.pas}
{$I cmd_mbmn.pas}
{$I cmd_mcro.pas}
{$I cmd_paus.pas}
{$I cmd_pclr.pas}
{$I cmd_pipe.pas}
{$I cmd_prnt.pas}
{$I cmd_prop.pas}
{$I cmd_read.pas}
{$I cmd_rnmt.pas}
{$I cmd_rst.pas}
{$I cmd_run.pas}
{$I cmd_scfg.pas}
{$I cmd_sdmt.pas}
{$I cmd_secn.pas}
{$I cmd_serd.pas}
{$I cmd_set.pas}
{$I cmd_sewr.pas}
{$I cmd_sreg.pas}
{$I cmd_srv.pas}
{$I cmd_sscr.pas}
{$I cmd_stck.pas}
{$I cmd_str.pas}
{$I cmd_sys.pas}
{$I cmd_tccn.pas}
{$I cmd_tcrd.pas}
{$I cmd_tcwr.pas}
{$I cmd_udcn.pas}
{$I cmd_udrd.pas}
{$I cmd_udwr.pas}
{$I cmd_var.pas}
{$I cmd_varr.pas}
{$I cmd_vrmn.pas}
{$I cmd_whvr.pas}
{$I cmd_wrte.pas}

{$I parsing.pas}
{$I fllprmpt.pas}

// SIMPLE COMMAND LINE
procedure simplecommandline;
var
  command: string;
  c: char;
begin
  if runmethod = 0 then writeln(PRGNAME + ' v' + PRGVERSION);
  repeat
    if runmethod = 3 then
    begin
      textbackground(uconfig.colors[1]);
      textcolor(uconfig.colors[0]);
    end;
    write(fullprompt);
    command := '';
    repeat
      if runmethod = 3 then showtime(colors[0], colors[1]);
      scheduler;
      delay(SHOWTIMEDELAY);
    until keypressed;
    repeat
      repeat
        if runmethod = 3 then showtime(colors[0], colors[1]);
        delay(SHOWTIMEDELAY);
      until keypressed;
      c := readkey;
      // detect hotkeys
      if c = #0 then
      begin
        repeat
          if runmethod = 3 then showtime(colors[0], colors[1]);
          delay(SHOWTIMEDELAY);
        until keypressed;
        c := readkey;
        // only insert
        if c = #46 then
          begin command := COMMANDS[17]; c := #32; end;                       // ALT-C
        if c = #18 then
          begin command := COMMANDS[15]; c := #32; end;                       // ALT-E
        if c = #23 then
          begin command := COMMANDS[22]; c := #32; end;                       // ALT-I
        if c = #34 then
          begin command := COMMANDS[2]; c := #32; end;                        // ALT-G
        if c = #38 then
          begin command := COMMANDS[4]; c := #32; end;                        // ALT-L
        if c = #50 then
          begin command := COMMANDS[88]; c := #32; end;                       // ALT-M
        if c = #25 then
          begin command := COMMANDS[5]; c := #32; end;                        // ALT-P
        if c = #19 then
          begin command := COMMANDS[6]; c := #32; end;                        // ALT-R
        if c = #20 then
          begin command := COMMANDS[7]; c := #32; end;                        // ALT-T
        if c = #31 then
          begin command := COMMANDS[8]; c := #32; end;                        // ALT-S
        if c = #17 then
          begin command := COMMANDS[11]; c := #32; end;                       // ALT-W
        if c = #59 then
          begin command := COMMANDS[3]; c:=#13; end;                          // F1
        if c = #60 then
          begin command := COMMANDS[13] + #32 + vars[12].vvalue; c:=#13; end; // F2
        if c = #61 then
          begin command := COMMANDS[14] + #32 + vars[12].vvalue; c:=#13; end; // F3
        if c = #62 then
          begin command := COMMANDS[18] + #32 + vars[12].vvalue; c:=#13; end; // F4
        if c = #63 then
          begin command := COMMANDS[19] + #32 + vars[12].vvalue; c:=#13; end; // F5
        if c = #64 then
          begin command := COMMANDS[33] + #32 + vars[12].vvalue; c:=#13; end; // F6
        if c = #65 then
          begin command := COMMANDS[35]; c:=#13; end;                         // F7
        if c = #66 then
          begin command := COMMANDS[12]; c:=#13; end;                         // F8
        if c = #67 then
          begin command := COMMANDS[38] + #32 + METHOD[3]; c:=#13; end;       // F9
        if c = #68 then
          begin command := COMMANDS[1]; c:=#13; end;                          // F10
        if c = #133 then
          begin command := COMMANDS[41]; c:=#13; end;                         // F11
        if c = #134 then
          begin command := COMMANDS[40]; c:=#13; end;                         // F12
        if c = #85 then
          begin command := COMMANDS[93]; c := #32; end;                       // SHIFT-F2
        if c = #86 then
          begin command := COMMANDS[39]; c := #32; end;                       // SHIFT-F3
        if c = #87 then
          begin command := COMMANDS[91]; c := #32; end;                       // SHIFT-F4
        if c = #90 then
          begin command := COMMANDS[112]; c := #32; end;                      // SHIFT-F7
        if c = #91 then
          begin command := COMMANDS[92]; c := #32; end;                       // SHIFT-F8
        if c = #92 then
          begin command := COMMANDS[119]; c := #32; end;                      // SHIFT-F9
        if c = #00 then
          begin command := COMMANDS[120]; c := #32; end;                      // CTRL-F9
        if c = #72 then                                                       // UP
          if uconfig.histitem > 0 then
          begin
            command := uconfig.histbuff[uconfig.histitem];
            dec(uconfig.histitem);
          end;
        if c = #80 then                                                       // DOWN
          if (uconfig.histitem < 255) and
             (length(uconfig.histbuff[uconfig.histitem + 1]) > 0) then
          begin
            inc(uconfig.histitem);
            command := uconfig.histbuff[uconfig.histitem];
          end;
      end else
      begin
        if c = #8 then delete(command, length(command), 1);                   // BACKSPACE
        if c = #9 then c := #32;                                              // TAB
        if c = #27 then command := '';                                        // ESC
        if (c <> #8) and (c <> #13) and (c <> #27) then command := command + c;
      end;
      xywrite(1, wherey, true, fullprompt + command);
    until (c = #13);                                                          // ENTER
    if length(command) > 0 then
    begin
      if uconfig.histlast < 255 then
      begin
        uconfig.histbuff[uconfig.histlast] := command;
        inc(uconfig.histlast);
      end else
      begin
        for b := 1 to 255 do
          uconfig.histbuff[b - 1] := uconfig.histbuff[b];
        uconfig.histbuff[uconfig.histlast] := command;
      end;
      uconfig.histitem := uconfig.histlast;
    end;
    {$IFNDEF X} writeln; {$ENDIF}
    parsingcommands(command);
    varmon_viewer;
  until command = COMMANDS[1]; // exit to OS
end;

// FULL SCREEN COMMAND LINE
procedure fullscreencommandline;
begin
  window(1, 1, termwidth, termheight);
  textbackground(lightgray); textcolor(black); clrscr;
  xywrite(1, 1, true, ' ' + PRGNAME + ' v' + PRGVERSION);
  xywrite((termwidth div 2) - (length(MSG02) div 2), 1, false, MSG02);
  gotoxy(2,termheight); ewrite(black, red, MSG01);
  window(1, 2, termwidth, termheight - 1);
  textbackground(uconfig.colors[1]); textcolor(uconfig.colors[0]); clrscr;
  window(1, 2, termwidth, termheight - 1);
  simplecommandline;
  window(1, 1, termwidth, termheight);
  textbackground(black); textcolor(lightgray); clrscr;
end;

{$I intrprtr.pas}

// - COMMAND LINE ARGUMENTS --------------------------------------------------
// SHOW USEABLE ARGUMENTS
procedure help(mode: boolean);
var
  b: byte;
begin
  if mode
    then writeln('There are one or more bad argument in command line.') else
  begin
    writeln('Usage: ' + BASENAME + ' [argument]');
    writeln;
    writeln('arguments:');
    for b := 0 to 3 do
    begin
      write('  ',CMDLINEPARAMS[b, 0]);
      gotoxy(8, wherey); write(CMDLINEPARAMS[b, 1]);
      gotoxy(30, wherey); writeln(CMDLINEPARAMS[b, 2]);
    end;
  end;
  quit(0, false, '');
end;

{$I version.pas}

// -- MAIN PROGRAM -------------------------------------------------------------
begin
  randomize;
  // detect language
  lang := getlang;
  translatemessages(LANG, BASENAME, '.mo');
  // check size of terminal
  if not terminalsize(MINTERMX, MINTERMY) then quit(1, false, ERR99);
  // parse command line arguments
  {$IFDEF WINDOWS}
    // load dlls
    loaded_inpout32dll := loadinpout32dll;
    if not loaded_inpout32dll then writeln(ERR98);
  {$ENDIF}
  runmethod := 0;
  if length(paramstr(1)) <> 0 then
  begin
    for b := 0 to 3 do
    begin
      if paramstr(1) = CMDLINEPARAMS[b, 0] then runmethod := b + 1;
      if paramstr(1) = CMDLINEPARAMS[b, 1] then runmethod := b + 1;
    end;
    case runmethod of
      1: help(false);
      2: version(true);
      0: help(true);
    end;
  end;
  // load configuration
  loadconfiguration(BASENAME, '.ini');
  vars[12].vvalue := uconfig.lastproject;
  // set temporary colors
  printcolors[0] := uconfig.colors[0];
  printcolors[1] := uconfig.colors[1];
  // set default constants and macros
  setdefaultconstants;
  setdefaultmacros;
  vars[11].vvalue := getuserdir;
  if length(vars[12].vvalue) = 0 then vars[12].vvalue := 'default';
  {$IFDEF GO32V2}
    vars[13].vvalue := getexedir + vars[12].vvalue;
  {$ELSE}
    vars[13].vvalue := vars[11].vvalue + PRGNAME + SLASH + vars[12].vvalue;
  {$ENDIF}
  // make, store and set directories
  ForceDirectories(vars[13].vvalue);
  if runmethod <> 4 then
  begin
    originaldirectory := getcurrentdir;
    setcurrentdir(vars[11].vvalue + PRGNAME);
  end;
  // pass command line arguments
  if runmethod = 4 then
  begin
    setdefaultvariables;
    cmd_const('ARGCNT', inttostr(paramcount - 2));
    cmd_const('ARG0', paramstr(2));
    if paramcount > 2 then
      for b := 3 to paramcount do
        cmd_const('ARG' + inttostr(b - 2), paramstr(b));
  end;
  // run main function
  case runmethod of
    0: simplecommandline;
    3: fullscreencommandline;
    4: interpreter(paramstr(2));
  end;
  // save configuration
  uconfig.lastproject := vars[12].vvalue;
  saveconfiguration(BASENAME, '.ini');
  {$IFDEF WINDOWS}
    // unload dlls
    if loaded_inpout32dll then unloadinpout32dll;
  {$ENDIF}
  // restore directory
  if runmethod <> 4 then setcurrentdir(originaldirectory);
  // set own or script's exit code
  if runmethod <> 4 then quit(0, false, '') else quit(scriptexitcode, false, '');
end.
